labrad/backend.py

```diff
@@ -65,15 +68,31 @@ def sendMessage(self, target, records, *args, **kw):
 
     class TwistedConnection(BaseConnection):
         def _connect(self, password, _timeout, tls_mode):
+            @defer.inlineCallbacks
+            def _connect_deferred():
+                cxn = yield getConnection(self.host, self.port, self.name,
+                                          password, tls_mode=tls_mode)
+                self._connected.set()
+
+                @defer.inlineCallbacks
+                def await_disconnect():
+                    try:
+                        yield cxn.onDisconnect()
+                    except Exception:
+                        pass
+                    self._connected.clear()
+                await_disconnect()
```

Comment:
In that case, maybe just set a callback on cxn.onDisconnect() explicitly,
rather than creating a new deferred that has no callbacks when it fires.

On Fri, Mar 4, 2016 at 1:53 PM, Matthew Neeley <notifications@github.com>
wrote:

> In labrad/backend.py
> <https://github.com/labrad/pylabrad/pull/226#discussion_r55094356>:
>
> > @@ -65,15 +68,31 @@ def sendMessage(self, target, records, *args, **kw):
> >
> >      class TwistedConnection(BaseConnection):
> >          def _connect(self, password, _timeout, tls_mode):
> > +            @defer.inlineCallbacks
> > +            def _connect_deferred():
> > +                cxn = yield getConnection(self.host, self.port, self.name,
> > +                                          password, tls_mode=tls_mode)
> > +                self._connected.set()
> > +
> > +                @defer.inlineCallbacks
> > +                def await_disconnect():
> > +                    try:
> > +                        yield cxn.onDisconnect()
> > +                    except Exception:
> > +                        pass
> > +                    self._connected.clear()
> > +                await_disconnect()
>
> I don't want to yield here because that would cause this to block until
> the connection is dropped, so we'd never actually return in the thread
> where connect was called. I will make a comment about this in the code;
> also, the name await_disconnect is probably misleading in this regard.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/labrad/pylabrad/pull/226/files#r55094356>.
>